### PR TITLE
Fixes #1086 memory leak in TabManager restore tabs

### LINKED ISSUE #1085: BVC taskQueue is refilled in setupTabs() leading to crash.
### Steps to Reproduce 
  1. Open Settings
  2. Select private browsing mode
  

**Actual result:** 
Crash

**Expected result:**
No crash. App moves to private browsing

**Reproduces how often:** [Easily reproduced, Intermittent Issue]

**Brave Version:** 

**Device details:** 

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)?

### LINKED ISSUE #1086: Memory leak in TabManager restoreAllTabs.
Restored tab is retained by the lazy stored property restoreAllTabs, leading to memory leak.

### Steps to Reproduce 
  1.
  2.
  3.

**Actual result:** 

**Expected result:**

**Reproduces how often:** [Easily reproduced, Intermittent Issue]

**Brave Version:** 

**Device details:** 

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)?

### PR DESCRIPTION
**Pull request by danishjafri88**
Fixes  #1086
1086 is fixed by making the implementation as function and managing once execution by a static variable.

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Reviewer Checklist:

- [x] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [x] Issues are assigned to at least one epic.
- [x] Issues include necessary QA labels:
  - [x] `QA/(Yes|No)`
  - [x] `release-notes/(include|exclude)`
  - [x] `bug` / `enhancement`
- [x] Necessary security reviews have taken place.
- [x] Adequate test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable)

### CODE REVIEW

File: Client/Frontend/Browser/TabManager.swift [hunk 1]
```diff
@@ -766,9 +766,14 @@ class TabManager: NSObject {
         imageStore?.clearExcluding(savedUUIDs)
     }
 
-    fileprivate lazy var restoreTabsInternal: () -> Tab? = {
+    fileprivate static var restoreOnceToken = false
```
Comment (jhreis) [inline Client/Frontend/Browser/TabManager.swift:769]: I think using `dispatch_once_t` is probably going to be better. It was designed for things like this, as opposed to a flat boolean.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 2]
```diff
@@ -2976,7 +2976,7 @@ class DeferredCountable<T> {
     }
     
     func fill(_ value: T) {
-        await.fill(value)
+        await.fillIfUnfilled(value)
```
Comment (jhreis) [inline Client/Frontend/Browser/BrowserViewController.swift]: This should either be included in a separate PR, since it solves a different case, or at the _very least_ a unique commit with its own message.

File: Client/Frontend/Browser/TabManager.swift [hunk 3]
```diff
@@ -836,15 +841,15 @@ class TabManager: NSObject {
         }
     }
 
-    lazy var restoreAllTabs: () -> Tab = {
+    func restoreAllTabs() -> Tab {
```
Comment (jhreis) [inline Client/Frontend/Browser/TabManager.swift:844]: Needs _run once_ protection too.

**Review by jhreis — CHANGES_REQUESTED** [review]
Should add some additional information to the tickets that were filed too, including STR. This is helpful to identify why something was changed earlier.

File: Client/Frontend/Browser/TabManager.swift (hunk 1 above)
Comment (danishjafri88) [reply Client/Frontend/Browser/TabManager.swift:769]: `dispatch_once_t` is no longer available in swift. The guide now promotes this method or lazy var for such purposes

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 2 above)
Comment (danishjafri88) [reply Client/Frontend/Browser/BrowserViewController.swift]: I will make a separate commit

File: Client/Frontend/Browser/TabManager.swift (hunk 3 above)
Comment (danishjafri88) [reply Client/Frontend/Browser/TabManager.swift:844]: This does not require a run once logic as the underlying logic `restoreTabsInternal ` is already run once, which on other case will return nil and here it will add a new tab. If we were to do a run once here, then too we have to create a new tab.

File: Client/Frontend/Browser/TabManager.swift [hunk 4]
```diff
@@ -766,9 +766,14 @@ class TabManager: NSObject {
         imageStore?.clearExcluding(savedUUIDs)
     }
 
-    fileprivate lazy var restoreTabsInternal: () -> Tab? = {
+    fileprivate static var restoreOnceToken = false
+    fileprivate func restoreTabsInternal() -> Tab? {
+        assert(Thread.isMainThread)
+        guard !TabManager.restoreOnceToken else { return nil }
```
Comment (jhreis) [inline Client/Frontend/Browser/TabManager.swift]: Double negative here is kind of weird, just use `if`:

```suggestion
        if TabManager.restoreOnceToken { return nil }
```

**Review by iccub — APPROVED** [review]

**Review by jhreis — CHANGES_REQUESTED** [review]
